src/mgos_imu_icm20948.c

```diff
@@ -0,0 +1,482 @@
+/*
+ * Copyright 2019 Google Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "mgos.h"
+#include "mgos_i2c.h"
+#include "mgos_imu_icm20948.h"
+
+#define ACC_BASE_ODR  1125.f
+#define GYRO_BASE_ODR 1100.f
+
+static bool mgos_imu_icm20948_change_bank(struct mgos_i2c *i2c, uint8_t i2caddr, void *imu_user_data, uint8_t bank_no) {
+  struct  mgos_imu_icm20948_userdata *iud = (struct mgos_imu_icm20948_userdata *)imu_user_data;
+  uint8_t bank_addr = 0x00;
+
+  if(bank_no == iud->current_bank_no) {
+    return true;
+  }
+
+  switch (bank_no) {
+    case 0: bank_addr = 0x00; break;
+    case 1: bank_addr = 0x10; break;
+    case 2: bank_addr = 0x20; break;
+    case 3: bank_addr = 0x30; break;
+  }
+
+  if(mgos_i2c_write_reg_b(i2c, i2caddr, MGOS_ICM20948_REG0_BANK_SEL, bank_addr)) {
+    iud->current_bank_no = bank_no;
+    return true;
+  }
+
+  return false;
+}
+
+static bool mgos_imu_icm20948_detect(struct mgos_i2c *i2c, uint8_t i2caddr, void *imu_user_data) {
+  int device_id;
+
+  if (!i2c) {
+    return false;
+  }
+
+  if(!mgos_imu_icm20948_change_bank(i2c, i2caddr, imu_user_data, 0)) {
```

Comment:
It's undesirable to detect a device by writing to it. What happens if the device isn't an ICM20948? You may corrupt its state or change its settings.
For a detection, it's sufficient to  read WHO_AM_I and compare to _DEVID which you do below. 

Then upon creation, you also set the bank to 0, so you should be able to remove it here.

Reply:
Yeah, you right, it is a bad practice.
Removed `*_change_bank`, but since `*_acc_create` mangle bank address, *gyro_detect that goes afterward does not work, so I placed `if(iud->accgyro_initialized) return true;` - since acc is created, that is definitely ICM20948.